symbolic-uplc/src/ZkFold/Symbolic/UPLC/Evaluation.hs

```diff
@@ -319,11 +397,144 @@ instance
+lessThanIntegerFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $
+                  fromConstant (\v w -> Symbolic.just @c (v Symbolic.< w))
+
+lessThanEqualsIntegerFun :: forall c .(Sym c) => Fun [BTInteger, BTInteger] BTBool c
+lessThanEqualsIntegerFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $
+                  fromConstant (\v w -> Symbolic.just @c (v Symbolic.<= w))
+
+--------------------------------------------------------------------------------
+
+appendByteStringFun :: forall c .(Sym c) => Fun [BTByteString, BTByteString] BTByteString c
+appendByteStringFun = fromConstant (\v w -> Symbolic.just @c $ dropZeros @BSLength (append v w))
+
+consByteStringFun :: forall c .(Sym c) => Fun [BTInteger, BTByteString] BTByteString c
+consByteStringFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $
+  fromConstant (\v w -> let bu :: ByteString 8 c = dropN @8 @IntLength (from $ uint v)
+                         in Symbolic.just @c $ dropZeros @BSLength (append w $ fromByteString bu))
+
+
+sliceByteStringFun :: forall c .(Sym c) => Fun [BTInteger, BTInteger, BTByteString] BTByteString c
+sliceByteStringFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $ withIsoFieldElemUInt @IntLength @c $
+  fromConstant (\(Int f :: Int IntLength Auto c) (Int t :: Int IntLength Auto c) (VarByteString l b)
+                    -> let (fr, to) = (Symbolic.max (from f) zero, Symbolic.max (from f + from t-one) (l - one))
+                           fe8 = fromConstant (8::Natural) :: FieldElement c
+                           newB = shiftR (shiftL b (fromConstant (value @BSLength) - l + fr*fe8)) (fromConstant (value @BSLength) - to*fe8)
+                        in fromConstant $ Symbolic.just @c (VarByteString to newB) :: (Fun '[] BTByteString c))
+
+isoFieldElemUInt :: forall n c. (Symbolic c) => Dict (NumberOfBits (BaseField c) ~ n)
+isoFieldElemUInt = unsafeAxiom
+
+withIsoFieldElemUInt :: forall n c {r}. (Symbolic c) => (NumberOfBits (BaseField c) ~ n => r) -> r
+withIsoFieldElemUInt =  withDict $ isoFieldElemUInt @n @c
+
+lengthOfByteStringFun :: forall c. (Sym c) => Fun '[BTByteString] BTInteger c
+lengthOfByteStringFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $ withIsoFieldElemUInt @IntLength @c $
+  fromConstant (\(VarByteString l _) -> Symbolic.just @c $ div (from (from l :: UInt IntLength Auto c)) (fromConstant (8 :: Natural)))
+
+indexByteStringFun :: forall c .(Sym c) => Fun [BTByteString, BTInteger] BTInteger c
+indexByteStringFun = withNumberOfRegisters @IntLength @Auto @(BaseField c) $ withIsoFieldElemUInt @IntLength @c $
+  withGetRegisterSize @IntLength @Auto @(BaseField c) $ withCeilRegSize @(GetRegisterSize (BaseField c) IntLength Auto) @OrdWord $
+    fromConstant (\(VarByteString l b) i@(Int t) ->
+      let fr = bool (error "index is not in interval") (from t) (isNotNegative i && t Symbolic.< fromConstant (value @BSLength))
```

Comment:
This is going to throw an error instantly when evaluated into circuit

Reply:
Instead of this, you have to return a `Symbolic.nothing` in case when an index is out of bounds